selfdrive/car/mazda/carstate.py

```diff
@@ -0,0 +1,214 @@
+from cereal import car
+from selfdrive.config import Conversions as CV
+from opendbc.can.parser import CANParser
+from selfdrive.car.interfaces import CarStateBase
+from selfdrive.car.mazda.values import DBC, LKAS_LIMITS, CAR
+
+GearShifter = car.CarState.GearShifter
+
+class STEER_LKAS():
+  def __init__(self):
+    self.block = 1
+    self.track = 1
+    self.handsoff = 0
+
+class CarState(CarStateBase):
+  def __init__(self, CP):
+    super().__init__(CP)
+
+    self.steer_lkas = STEER_LKAS()
+
+    self.acc_active_last = False
+    self.speed_kph = 0
+    self.lkas_speed_lock = False
+    self.low_speed_lockout = True
+    self.low_speed_lockout_last = True
+    self.acc_press_update = False
+
+  def update(self, cp, cp_cam):
+
+    ret = car.CarState.new_message()
+    ret.wheelSpeeds.fl = cp.vl["WHEEL_SPEEDS"]['FL'] * CV.KPH_TO_MS
+    ret.wheelSpeeds.fr = cp.vl["WHEEL_SPEEDS"]['FR'] * CV.KPH_TO_MS
+    ret.wheelSpeeds.rl = cp.vl["WHEEL_SPEEDS"]['RL'] * CV.KPH_TO_MS
+    ret.wheelSpeeds.rr = cp.vl["WHEEL_SPEEDS"]['RR'] * CV.KPH_TO_MS
+    ret.vEgoRaw = (ret.wheelSpeeds.fl + ret.wheelSpeeds.fr + ret.wheelSpeeds.rl + ret.wheelSpeeds.rr) / 4.
+    ret.vEgo, ret.aEgo = self.update_speed_kf(ret.vEgoRaw)
+    ret.standstill = ret.vEgoRaw < 0.01
+
+    self.speed_kph =  ret.vEgoRaw  // CV.KPH_TO_MS
+
+    gear = int(cp.vl["GEAR"]['GEAR']) >> 1
+
+    if gear >= 1 & gear <= 5:
+      ret.gearShifter = GearShifter.drive
+    elif gear == 0:
+      ret.gearShifter = GearShifter.park
+    elif gear == 6:
+      ret.gearShifter = GearShifter.reverse
+    elif gear == 7:
+      ret.gearShifter = GearShifter.neutral
+    else:
+      ret.gearShifter = GearShifter.unknown
```

Comment:
Why the shift for the gear value? Also seems like a pretty big range for drive. If you define these values in the DBC, you can use [parse_gear_shifter](https://github.com/commaai/openpilot/blob/master/selfdrive/car/interfaces.py#L158) like the other cars.

Reply:
Shift because I found that the low bit is not used, i.e shouldn't be part of the signal. I already opened a PR to fix that (https://github.com/commaai/opendbc/pull/254) once that is merged the shift can be dropped. 1 through 5 are gear positions for speeds  1 to 5.  I did see this that function, do I fill in letters instead and leave the mapping to the function?

Reply:
You have to define the gear values in the DBC like this https://github.com/commaai/opendbc/blob/master/toyota_prius_2017_pt_generated.dbc#L401

Reply:
Do I edit the file manually or can this be done via cabana?

Reply:
I don't think cabana can do this. I'd just do it manually

Reply:
here is what I came up with: 
```
VAL_ 552 GEAR 0 "P" 1 "D" 2 "D" 3 "D" 4 "D" 5 "D" 6 "R" 7 "N";
```
There might be a better signal to  track the "stick" state, but I haven't found that one if it does exist.

Reply:
Does this work?

Reply:
Yeah, that should work. What do you mean by stick state? If you mean the actual gear position, I'd just put it in a comment